**Objective:**
* Practice Naive Bayes algorithm based classification.•Identify the predictors that can be of influence by experiment.

1. Load the kinematics dataset as measured on mobile sensors from the file “run_or_walk.csv”. List out the columns in the dataset.

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing

#Visualisation frameworks
import matplotlib.pyplot as plt
%matplotlib inline  
import seaborn as sns

In [16]:
df = pd.read_csv('C:/Users/gabri/Desktop/Data Science/Python/Python for Data Science - Edureka/Module 9 - Supervised Learning 2/run_or_walk.csv')
print(df.shape)
df.head()

(88588, 11)


date                time username  wrist  activity  acceleration_x  \
0  2017-6-30  13:51:15:847724020   viktor      0         0          0.2650   
1  2017-6-30  13:51:16:246945023   viktor      0         0          0.6722   
2  2017-6-30  13:51:16:446233987   viktor      0         0          0.4399   
3  2017-6-30  13:51:16:646117985   viktor      0         0          0.3031   
4  2017-6-30  13:51:16:846738994   viktor      0         0          0.4814   

   acceleration_y  acceleration_z  gyro_x  gyro_y  gyro_z  
0         -0.7814         -0.0076 -0.0590  0.0325 -2.9296  
1         -1.1233         -0.2344 -0.1757  0.0208  0.1269  
2         -1.4817          0.0722 -0.9105  0.1063 -2.4367  
3         -0.8125          0.0888  0.1199 -0.4099 -2.9336  
4         -0.9312          0.0359  0.0527  0.4379  2.4922

In [5]:
df.columns

Index(['date', 'time', 'username', 'wrist', 'activity', 'acceleration_x',
       'acceleration_y', 'acceleration_z', 'gyro_x', 'gyro_y', 'gyro_z'],
      dtype='object')

2. Let the target variable ‘y’ be the activity and assign all the columns after it to ‘x’.

In [38]:
df.activity.unique()

array([0, 1], dtype=int64)

In [39]:
df.username.unique()

array(['viktor'], dtype=object)

In [26]:
X = df.drop(['activity','date','time','username'], axis=1)
y = df['activity']

X.head()

wrist  acceleration_x  acceleration_y  acceleration_z  gyro_x  gyro_y  \
0      0          0.2650         -0.7814         -0.0076 -0.0590  0.0325   
1      0          0.6722         -1.1233         -0.2344 -0.1757  0.0208   
2      0          0.4399         -1.4817          0.0722 -0.9105  0.1063   
3      0          0.3031         -0.8125          0.0888  0.1199 -0.4099   
4      0          0.4814         -0.9312          0.0359  0.0527  0.4379   

   gyro_z  
0 -2.9296  
1  0.1269  
2 -2.4367  
3 -2.9336  
4  2.4922

In [27]:
y.head()

0    0
1    0
2    0
3    0
4    0
Name: activity, dtype: int64

3. Using Scikit-learn fit a Gaussian Naive Bayes model and observe the accuracy. Generate a classification report using scikitlearn.

In [34]:
#Tools
from sklearn.model_selection import train_test_split
from sklearn import metrics

#Model
from sklearn.naive_bayes import GaussianNB

In [29]:
# Split into train/test
X_train, X_test, y_train, y_test = \
    train_test_split(X, y, test_size=0.2, stratify=y,
                     random_state=0)

print('X_train shape:', X_train.shape, '\nX_test shape:', X_test.shape)

X_train shape: (70870, 7) 
X_test shape: (17718, 7)


In [36]:
model = GaussianNB()
model.fit(X_test, y_test)

# Predict the value of the digit on the test subset
predicted = model.predict(X_test)

In [37]:
print(f"Classification report for classifier {GaussianNB}:\n"
      f"{metrics.classification_report(y_test, predicted)}\n")

Classification report for classifier <class 'sklearn.naive_bayes.GaussianNB'>:
              precision    recall  f1-score   support

           0       0.93      0.99      0.96      8845
           1       0.99      0.93      0.96      8873

    accuracy                           0.96     17718
   macro avg       0.96      0.96      0.96     17718
weighted avg       0.96      0.96      0.96     17718




4. Repeat the model once using only the acceleration values as predictors and then using only the gyro values as predictors. Comment on the difference in accuracy between both the models.

In [40]:
df.head()

date                time username  wrist  activity  acceleration_x  \
0  2017-6-30  13:51:15:847724020   viktor      0         0          0.2650   
1  2017-6-30  13:51:16:246945023   viktor      0         0          0.6722   
2  2017-6-30  13:51:16:446233987   viktor      0         0          0.4399   
3  2017-6-30  13:51:16:646117985   viktor      0         0          0.3031   
4  2017-6-30  13:51:16:846738994   viktor      0         0          0.4814   

   acceleration_y  acceleration_z  gyro_x  gyro_y  gyro_z  
0         -0.7814         -0.0076 -0.0590  0.0325 -2.9296  
1         -1.1233         -0.2344 -0.1757  0.0208  0.1269  
2         -1.4817          0.0722 -0.9105  0.1063 -2.4367  
3         -0.8125          0.0888  0.1199 -0.4099 -2.9336  
4         -0.9312          0.0359  0.0527  0.4379  2.4922

### Acceleration only

In [42]:
X = df.drop(['activity','date','time','username', 'wrist','gyro_x',	'gyro_y', 'gyro_z'], axis=1)

X.head()

acceleration_x  acceleration_y  acceleration_z
0          0.2650         -0.7814         -0.0076
1          0.6722         -1.1233         -0.2344
2          0.4399         -1.4817          0.0722
3          0.3031         -0.8125          0.0888
4          0.4814         -0.9312          0.0359

In [43]:
# Split into train/test
X_train, X_test, y_train, y_test = \
    train_test_split(X, y, test_size=0.2, stratify=y,
                     random_state=0)

print('X_train shape:', X_train.shape, '\nX_test shape:', X_test.shape)

X_train shape: (70870, 3) 
X_test shape: (17718, 3)


In [44]:
model = GaussianNB()
model.fit(X_test, y_test)

# Predict the value of the digit on the test subset
predicted = model.predict(X_test)

print(f"Classification report for classifier {GaussianNB}:\n"
      f"{metrics.classification_report(y_test, predicted)}\n")

Classification report for classifier <class 'sklearn.naive_bayes.GaussianNB'>:
              precision    recall  f1-score   support

           0       0.93      0.99      0.96      8845
           1       0.99      0.92      0.96      8873

    accuracy                           0.96     17718
   macro avg       0.96      0.96      0.96     17718
weighted avg       0.96      0.96      0.96     17718




### Gyro only

In [45]:
X = df.drop(['activity','date','time','username', 'wrist','acceleration_x',	'acceleration_y', 'acceleration_z'], axis=1)

X.head()

gyro_x  gyro_y  gyro_z
0 -0.0590  0.0325 -2.9296
1 -0.1757  0.0208  0.1269
2 -0.9105  0.1063 -2.4367
3  0.1199 -0.4099 -2.9336
4  0.0527  0.4379  2.4922

In [46]:
# Split into train/test
X_train, X_test, y_train, y_test = \
    train_test_split(X, y, test_size=0.2, stratify=y,
                     random_state=0)

print('X_train shape:', X_train.shape, '\nX_test shape:', X_test.shape)

X_train shape: (70870, 3) 
X_test shape: (17718, 3)


In [47]:
model = GaussianNB()
model.fit(X_test, y_test)

# Predict the value of the digit on the test subset
predicted = model.predict(X_test)

print(f"Classification report for classifier {GaussianNB}:\n"
      f"{metrics.classification_report(y_test, predicted)}\n")

Classification report for classifier <class 'sklearn.naive_bayes.GaussianNB'>:
              precision    recall  f1-score   support

           0       0.62      0.74      0.67      8845
           1       0.68      0.55      0.61      8873

    accuracy                           0.65     17718
   macro avg       0.65      0.65      0.64     17718
weighted avg       0.65      0.65      0.64     17718




**Acceleration-only results (AVG):**
* precision: 0.96
* recall: 0.96
* f1-score: 0.96

**Gyro-only results (AVG):**
* precision: 0.65
* recall: 0.65
* f1-score: 0.64
